compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -168,10 +168,22 @@ impl<'rt, 'mir, 'tcx: 'mir, M: CompileTimeMachine<'mir, 'tcx, const_eval::Memory
         mplace: &MPlaceTy<'tcx>,
         fields: impl Iterator<Item = InterpResult<'tcx, Self::V>>,
     ) -> InterpResult<'tcx> {
-        // ZSTs cannot contain pointers, so we can skip them.
-        if mplace.layout.is_zst() {
+        // We want to walk the aggregate to look for reference types to intern. While doing that we
+        // also need to take special care of interior mutability.
+        //
+        // As an optimization, however, if the allocation does not contain any references: we don't
+        // need to do the walk. It can be costly for big arrays for example (e.g. issue #93215).
+
+        let Some((size, align)) = self.ecx.size_and_align_of_mplace(&mplace)? else {
+            // We could be dealing with an extern type here in the future, so we do the regular
+            // walk.
+            return self.walk_aggregate(mplace, fields);
```

Comment:
We are skipping the `UnsafeCell` logic here. Please add a comment explaining why that is okay.

Reply:
I've done this when addressing the next comment's points, with the refactor you suggested.